File: client/src/services/userManagementService.ts
import axios from 'axios';
import { 
    UserProfile, 
    UserCreationRequest, 
    UserUpdateRequest, 
    UserSummary, 
    EmergencyContactData
} from '../types/user';
import { authService } from './authService';

class UserManagementService {
    private baseUrl: string;

    constructor() {
        this.baseUrl = process.env.REACT_APP_USER_MANAGEMENT_SERVICE_URL || 'http://localhost:8082/api';
    }

    private getAuthHeader() {
        const token = localStorage.getItem('token');
        if (!token) {
            console.error('No auth token found in localStorage');
            throw new Error('Authentication required');
        }
        console.log('Using token:', token.substring(0, 10) + '...'); // Log first 10 chars of token
        return {
            headers: {
                Authorization: `Bearer ${token}`,
                'Content-Type': 'application/json'
            }
        };
    }

    async getUserProfile(userId: string): Promise<UserCreationRequest | null> {
        try {
            const response = await axios.get(`${this.baseUrl}/users/${userId}/profile`, this.getAuthHeader());
            return response.data;
        } catch (error) {
            if (axios.isAxiosError(error)) {
                console.error('Error response:', error.response?.data);
                if (error.response?.status === 404) {
                    return null;
                }
                if (error.response?.status === 403) {
                    throw new Error('Authentication failed. Please log in again.');
                }
            }
            console.error('Error fetching user profile:', error);
            throw error;
        }
    }

    async createUserProfile(data: UserCreationRequest): Promise<void> {
        try {
            console.log('Creating user profile with data:', data);
            const response = await axios.post(`${this.baseUrl}/users`, data, this.getAuthHeader());
            console.log('Profile creation response:', response.data);
        } catch (error) {
            if (axios.isAxiosError(error)) {
                console.error('Error response:', error.response?.data);
                if (error.response?.status === 403) {
                    throw new Error('Authentication failed. Please log in again.');
                }
            }
            console.error('Error creating user profile:', error);
            throw error;
        }
    }

    async updateUserProfile(data: UserUpdateRequest): Promise<void> {
        try {
            const currentUser = await authService.getCurrentUser();
            console.log('Updating profile for user:', currentUser.id);
            
            // Ensure emergency contacts are properly formatted
            const formattedData = {
                ...data,
                emergencyContacts: data.emergencyContacts.map(contact => ({
                    name: contact.name,
                    email: contact.email,
                    phone: contact.phone,
                    preferredMethod: contact.preferredMethod
                })),
                preferences: {
                    checkInInterval: data.preferences.checkInInterval,
                    shareLocation: data.preferences.shareLocation,
                    notifyOnDelay: data.preferences.notifyOnDelay,
                    autoNotifyContacts: data.preferences.autoNotifyContacts,
                    enableSOS: data.preferences.enableSOS
                }
            };

            console.log('Formatted update data:', JSON.stringify(formattedData, null, 2));
            
            const response = await axios.put(
                `${this.baseUrl}/users/${currentUser.id}`,
                formattedData,
                {
                    ...this.getAuthHeader(),
                    headers: {
                        ...this.getAuthHeader().headers,
                        'Content-Type': 'application/json'
                    }
                }
            );
            
            console.log('Update response:', response.data);

            // Verify the update was successful
            if (!response.data) {
                throw new Error('No response data received from server');
            }

            // Fetch the updated profile to verify changes
            const updatedProfile = await this.getUserProfile(currentUser.id);
            if (updatedProfile) {
                console.log('Updated profile:', updatedProfile);
                if (!updatedProfile.emergencyContacts || updatedProfile.emergencyContacts.length === 0) {
                    console.warn('Emergency contacts not saved in the response');
                }
            } else {
                console.warn('Could not fetch updated profile');
            }
        } catch (error) {
            console.error('Error updating user profile:', error);
            if (axios.isAxiosError(error)) {
                console.error('Response data:', error.response?.data);
                console.error('Response status:', error.response?.status);
                console.error('Request data:', error.config?.data);
            }
            throw error;
        }
    }

    async deleteUserProfile(userId: string): Promise<void> {
        await axios.delete(`${this.baseUrl}/users/${userId}`, this.getAuthHeader());
    }

    async addEmergencyContact(userId: string, contactUserCode: string): Promise<{ id: string; requesterId: string; contactUserId: string }> {
        const response = await axios.post(
            `${this.baseUrl}/users/${userId}/emergency-contacts?contactUserCode=${contactUserCode}`,
            {},
            this.getAuthHeader()
        );
        return response.data;
    }

    async getPendingEmergencyContacts(userId: string): Promise<EmergencyContactData[]> {
        const response = await axios.get(`${this.baseUrl}/users/${userId}/emergency-contacts/pending`, this.getAuthHeader());
        return response.data;
    }

    async respondToEmergencyContact(userId: string, requestId: string, accept: boolean): Promise<void> {
        await axios.put(
            `${this.baseUrl}/users/${userId}/emergency-contacts/${requestId}?accept=${accept}`,
            {},
            this.getAuthHeader()
        );
    }

    async getEmergencyContacts(userId: string): Promise<EmergencyContactData[]> {
        const response = await axios.get(`${this.baseUrl}/users/${userId}/emergency-contacts`, this.getAuthHeader());
        return response.data;
    }

    async getEmergencyContactsOf(userId: string): Promise<EmergencyContactData[]> {
        const response = await axios.get(`${this.baseUrl}/users/${userId}/emergency-contacts-of`, this.getAuthHeader());
        return response.data;
    }

    async removeEmergencyContact(userId: string, contactId: string): Promise<void> {
        await axios.delete(`${this.baseUrl}/users/${userId}/emergency-contacts/${contactId}`, this.getAuthHeader());
    }

    async removeEmergencyContactOf(userId: string, requesterId: string): Promise<void> {
        await axios.delete(`${this.baseUrl}/users/${userId}/emergency-contacts-of/${requesterId}`, this.getAuthHeader());
    }
}

export const userManagementService = new UserManagementService(); 

File: server/usermanagement-service/src/main/java/com/usermanagement_service/service/UserManagementService.java
package com.usermanagement_service.service;

import com.usermanagement_service.dto.*;
import com.usermanagement_service.model.*;
import com.usermanagement_service.repository.EmergencyContactRepository;
import com.usermanagement_service.repository.UserProfileRepository;
import com.usermanagement_service.repository.AuthUserRepository;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.HashSet;
import java.util.Set;
import java.util.ArrayList;

@Service
@RequiredArgsConstructor
public class UserManagementService {
    private final EmergencyContactRepository emergencyContactRepository;
    private final UserProfileRepository userProfileRepository;
    private final AuthUserRepository authUserRepository;
    private final UserCodeGenerator userCodeGenerator;
    private final ProfilePictureGenerator profilePictureGenerator;
    private static final Logger logger = LoggerFactory.getLogger(UserManagementService.class);

    public UserProfileResponse getUserProfile(String userId) {
        logger.info("Getting user profile for userId: {}", userId);
        
        // First check if user exists in auth service
        AuthUser authUser = authUserRepository.findById(userId)
            .orElseThrow(() -> {
                logger.error("User not found in auth service for userId: {}", userId);
                return new RuntimeException("User not found in auth service");
            });
            
        // Then check if profile exists
        com.usermanagement_service.model.UserProfile profile = userProfileRepository.findByUserId(userId)
            .orElse(null);
            
        // If no profile exists, return just the auth user data
        if (profile == null) {
            logger.info("No profile found for userId: {}, returning auth user data only", userId);
            return UserProfileResponse.builder()
                .userId(authUser.getId())
                .email(authUser.getEmail())
                .emergencyContacts(List.of()) // Return empty list for new profiles
                .build();
        }
        
        // Get emergency contacts
        List<EmergencyContact> emergencyContacts = emergencyContactRepository.findByRequesterIdAndStatus(userId, RequestStatus.ACCEPTED);
        List<EmergencyContactDTO> emergencyContactDTOs = emergencyContacts.stream()
            .map(contact -> EmergencyContactDTO.builder()
                .name(contact.getName())
                .email(contact.getEmail())
                .phone(contact.getPhone())
                .preferredMethod(contact.getPreferredMethod())
                .build())
            .collect(Collectors.toList());
        
        // If profile exists, merge the data
        return UserProfileResponse.builder()
            .id(profile.getId())
            .userId(profile.getUserId())
            .email(authUser.getEmail())
            .alias(profile.getAlias())
            .gender(profile.getGender())
            .ageGroup(profile.getAgeGroup())
            .preferences(profile.getPreferences() != null ? profile.getPreferences().getPreferences() : null)
            .preferredContactMethod(profile.getPreferredContactMethod())
            .phoneNr(profile.getPhoneNr())
            .profilePictureUrl(profile.getProfilePictureUrl())
            .emergencyContacts(emergencyContactDTOs)
            .build();
    }

    public UserProfileResponse createUserProfile(UserCreationRequest request) {
        logger.info("Creating user profile with request: {}", request);
        
        // First verify user exists in auth service
        AuthUser authUser = authUserRepository.findById(request.getId())
            .orElseThrow(() -> new RuntimeException("User not found in auth service"));
        
        // Check if profile already exists
        if (userProfileRepository.findByUserId(request.getId()).isPresent()) {
            throw new RuntimeException("Profile already exists for this user");
        }

        Map<String, Object> reqPrefs = request.getPreferences();
        Preferences preferences = Preferences.builder()
            .shareLocation(Boolean.TRUE.equals(reqPrefs.get("shareLocation")))
            .notifyOnDelay(Boolean.TRUE.equals(reqPrefs.get("notifyOnDelay")))
            .autoNotifyContacts(Boolean.TRUE.equals(reqPrefs.get("autoNotifyContacts")))
            .checkInInterval(reqPrefs.get("checkInInterval") != null ? ((Number) reqPrefs.get("checkInInterval")).intValue() : 30)
            .enableSOS(Boolean.TRUE.equals(reqPrefs.get("enableSOS")))
            .preferences(reqPrefs)
            .build();

        com.usermanagement_service.model.UserProfile profile = com.usermanagement_service.model.UserProfile.builder()
            .userId(request.getId())
            .alias(request.getAlias())
            .gender(request.getGender())
            .ageGroup(request.getAgeGroup())
            .preferences(preferences)
            .preferredContactMethod(request.getPreferredContactMethod())
            .phoneNr(request.getPhoneNr())
            .profilePictureUrl(request.getProfilePictureUrl())
            .build();
            
        profile = userProfileRepository.save(profile);
        logger.info("User profile created successfully for userId: {}", request.getId());
        
        return UserProfileResponse.builder()
            .id(profile.getId())
            .userId(profile.getUserId())
            .email(authUser.getEmail())
            .alias(profile.getAlias())
            .gender(profile.getGender())
            .ageGroup(profile.getAgeGroup())
            .preferences(profile.getPreferences() != null ? profile.getPreferences().getPreferences() : null)
            .preferredContactMethod(profile.getPreferredContactMethod())
            .phoneNr(profile.getPhoneNr())
            .profilePictureUrl(profile.getProfilePictureUrl())
            .emergencyContacts(List.of()) // Return empty list for new profiles
            .build();
    }

    @Transactional
    public UserProfileResponse updateUserProfile(String userId, UserUpdateRequest request) {
        logger.info("Updating user profile for userId: {}", userId);
        com.usermanagement_service.model.UserProfile profile = userProfileRepository.findByUserId(userId)
            .orElseThrow(() -> new RuntimeException("User profile not found"));

        // Update basic profile information
        profile.setAlias(request.alias());
        profile.setGender(request.gender());
        profile.setAgeGroup(request.ageGroup());
        profile.setPreferredContactMethod(request.preferredContactMethod());
        profile.setPhoneNr(request.phoneNr());

        // Update preferences
        Preferences reqPrefs = request.preferences();
        Preferences preferences = Preferences.builder()
            .shareLocation(reqPrefs.isShareLocation())
            .notifyOnDelay(reqPrefs.isNotifyOnDelay())
            .autoNotifyContacts(reqPrefs.isAutoNotifyContacts())
            .checkInInterval(reqPrefs.getCheckInInterval())
            .enableSOS(reqPrefs.isEnableSOS())
            .preferences(reqPrefs.getPreferences())
            .build();
        profile.setPreferences(preferences);

        // Handle emergency contacts
        if (request.emergencyContacts() != null && !request.emergencyContacts().isEmpty()) {
            // Delete existing emergency contacts
            emergencyContactRepository.deleteByRequesterId(userId);

            // Create and save new emergency contacts
            List<EmergencyContact> emergencyContacts = request.emergencyContacts().stream()
                .map(dto -> EmergencyContact.builder()
                    .requesterId(userId)
                    .contactUserId(dto.getEmail())
                    .name(dto.getName())
                    .email(dto.getEmail())
                    .phone(dto.getPhone())
                    .preferredMethod(dto.getPreferredMethod())
                    .status(RequestStatus.ACCEPTED)
                    .build())
                .collect(Collectors.toList());

            emergencyContacts = emergencyContactRepository.saveAll(emergencyContacts);
            logger.info("Saved {} emergency contacts for user {}", emergencyContacts.size(), userId);
        }

        // Save the updated profile
        profile = userProfileRepository.save(profile);
        logger.info("User profile updated successfully for userId: {}", userId);

        // Get the updated emergency contacts
        List<EmergencyContact> updatedEmergencyContacts = emergencyContactRepository.findByRequesterIdAndStatus(userId, RequestStatus.ACCEPTED);
        List<EmergencyContactDTO> updatedEmergencyContactDTOs = updatedEmergencyContacts.stream()
            .map(contact -> EmergencyContactDTO.builder()
                .name(contact.getName())
                .email(contact.getEmail())
                .phone(contact.getPhone())
                .preferredMethod(contact.getPreferredMethod())
                .build())
            .collect(Collectors.toList());

        // Build the response
        return UserProfileResponse.builder()
            .id(profile.getId())
            .userId(profile.getUserId())
            .alias(profile.getAlias())
            .gender(profile.getGender())
            .ageGroup(profile.getAgeGroup())
            .preferences(profile.getPreferences() != null ? profile.getPreferences().getPreferences() : null)
            .preferredContactMethod(profile.getPreferredContactMethod())
            .phoneNr(profile.getPhoneNr())
            .profilePictureUrl(profile.getProfilePictureUrl())
            .emergencyContacts(updatedEmergencyContactDTOs)
            .build();
    }

    public AddEmergencyContactResponse addEmergencyContact(String requesterId, String contactUserId) {
        com.usermanagement_service.model.UserProfile contactProfile = userProfileRepository.findByUserId(contactUserId)
            .orElseThrow(() -> new RuntimeException("Contact user not found"));

        if (emergencyContactRepository.existsByRequesterIdAndContactUserId(requesterId, contactProfile.getUserId())) {
            throw new RuntimeException("Emergency contact already exists");
        }

        EmergencyContact contact = EmergencyContact.builder()
            .requesterId(requesterId)
            .contactUserId(contactProfile.getUserId())
            .status(RequestStatus.PENDING)
            .build();

        contact = emergencyContactRepository.save(contact);
        return new AddEmergencyContactResponse(contact.getId(), contact.getRequesterId(), contact.getContactUserId());
    }

    public List<EmergencyContact> getPendingEmergencyContactRequests(String userId) {
        return emergencyContactRepository.findByContactUserIdAndStatus(userId, RequestStatus.PENDING);
    }

    public void respondToEmergencyContactRequest(String requestId, boolean accept) {
        EmergencyContact contact = emergencyContactRepository.findById(requestId)
            .orElseThrow(() -> new RuntimeException("Emergency contact request not found"));

        contact.setStatus(accept ? RequestStatus.ACCEPTED : RequestStatus.DENIED);
        emergencyContactRepository.save(contact);
    }

    public List<EmergencyContact> getEmergencyContacts(String userId) {
        return emergencyContactRepository.findByRequesterIdAndStatus(userId, RequestStatus.ACCEPTED);
    }

    public List<EmergencyContact> getEmergencyContactsOf(String userId) {
        return emergencyContactRepository.findByContactUserIdAndStatus(userId, RequestStatus.ACCEPTED);
    }

    public List<EmergencyContact> getEmergencyContactsFor(String userId) {
        return emergencyContactRepository.findByRequesterIdAndStatus(userId, RequestStatus.ACCEPTED);
    }

    public List<EmergencyContact> getEmergencyContactsOfFor(String userId) {
        return emergencyContactRepository.findByContactUserIdAndStatus(userId, RequestStatus.ACCEPTED);
    }

    public void removeEmergencyContact(String userId, String contactId) {
        emergencyContactRepository.deleteByRequesterIdAndContactUserId(userId, contactId);
    }

    public void removeEmergencyContactOf(String requesterId, String userId) {
        emergencyContactRepository.deleteByRequesterIdAndContactUserId(requesterId, userId);
    }

    public List<UserSearchResponse> searchUsers(String query, String currentUserId) {
        logger.info("Searching users with query: {} for user: {}", query, currentUserId);
        
        // Search by email or alias (case-insensitive)
        List<AuthUser> authUsers = authUserRepository.findByEmailContainingIgnoreCaseOrIdContainingIgnoreCase(query, query);
        List<com.usermanagement_service.model.UserProfile> profiles = userProfileRepository.findByAliasContainingIgnoreCase(query);
        
        // Combine and deduplicate results
        Set<String> foundUserIds = new HashSet<>();
        List<UserSearchResponse> results = new ArrayList<>();
        
        // Add auth users
        for (AuthUser authUser : authUsers) {
            if (!authUser.getId().equals(currentUserId) && !foundUserIds.contains(authUser.getId())) {
                foundUserIds.add(authUser.getId());
                
                // Get profile for alias
                com.usermanagement_service.model.UserProfile profile = userProfileRepository.findByUserId(authUser.getId()).orElse(null);
                
                results.add(UserSearchResponse.builder()
                    .userId(authUser.getId())
                    .email(authUser.getEmail())
                    .alias(profile != null ? profile.getAlias() : null)
                    .profilePictureUrl(profile != null ? profile.getProfilePictureUrl() : null)
                    .build());
            }
        }
        
        // Add profiles not already included
        for (com.usermanagement_service.model.UserProfile profile : profiles) {
            if (!profile.getUserId().equals(currentUserId) && !foundUserIds.contains(profile.getUserId())) {
                foundUserIds.add(profile.getUserId());
                
                // Get auth user for email
                AuthUser authUser = authUserRepository.findById(profile.getUserId()).orElse(null);
                
                results.add(UserSearchResponse.builder()
                    .userId(profile.getUserId())
                    .email(authUser != null ? authUser.getEmail() : null)
                    .alias(profile.getAlias())
                    .profilePictureUrl(profile.getProfilePictureUrl())
                    .build());
            }
        }
        
        // Limit results to 10
        return results.stream().limit(10).collect(Collectors.toList());
    }
} 